File: polyominoes.js
const transformations = [
  ([ x, y ]) => [  x,  y ], // identity
  ([ x, y ]) => [ -y,  x ], // rotate pi/2
  ([ x, y ]) => [ -x, -y ], // rotate pi
  ([ x, y ]) => [  y, -x ], // rotate 3pi/2
  ([ x, y ]) => [ -x,  y ], // reflect in x axis
  ([ x, y ]) => [  x, -y ], // reflect in y axis
  ([ x, y ]) => [  y,  x ], // reflect in x = y
  ([ x, y ]) => [ -y, -x ]  // reflect in x = -y
];

class Polyomino {
  // Reduces a representation of a polyomino to a string such that two
  // representations of the same polyomino (up to translation) will produce the
  // same string
  static normalize(representation) {
    const dx = Math.min(...representation.map(([ x, y ]) => x));
    const dy = Math.min(...representation.map(([ x, y ]) => y));
    // Make the co-ordinates as small as possible, while all still positive
    const normalized = representation.map(([ x, y ]) => [ x - dx, y - dy ])
    return JSON.stringify(normalized.sort());
  }

  constructor(representation) {
    // Store a set of the normalized representations of each rotation and
    // reflection of the polyomino
    this.representations = new Set(
      transformations.map(t => representation.map(square => t(square)))
        .map(Polyomino.normalize)
    );
  }

  // Tests if the polyomino is represented by the provided list of squares
  isRepresentedBy(representation) {
    return this.representations.has(Polyomino.normalize(representation));
  }

  // Gets an arbitrary represenation of the polyomino
  get representation() {
    return JSON.parse(this.representations.values().next().value);
  }
}

const directions = [ [ 1, 0 ], [ 0, 1 ], [ -1, 0 ], [ 0, -1 ] ];

function containsSquare(representation, square) {
  return representation.some(s => JSON.stringify(s) === JSON.stringify(square));
}

// Returns an iterable of representations that can be built by adding one
// square to the given representation
function* grow(representation) {
  for ([ x, y ] of representation) {
    for ([ dx, dy ] of directions) {
      if (!containsSquare(representation, [ x + dx, y + dy ])) {
        yield [ ...representation, [ x + dx, y + dy ] ];
      }
    }
  }
}

// Tests if the provided representation corresponds to one of the provided
// polyominoes
function alreadyRepresented(polyominoes, representation) {
  return polyominoes.some(p => p.isRepresentedBy(representation));
}

// Returns an array of all free n-ominoes
function polyominoes(n) {
  const ominoes = { 1: [ new Polyomino([ [ 0, 0 ] ]) ] };
  for (let i = 2; i <= n; i++) {
    ominoes[i] = [];
    for (omino of ominoes[i - 1]) {
      for (representation of grow(omino.representation)) {
        if (!alreadyRepresented(ominoes[i], representation)) {
          ominoes[i].push(new Polyomino(representation));
        }
      }
    }
  }
  return ominoes[n];
}

module.exports = polyominoes;

// # Usage examples
//
// Find out how many 8-ominoes there are:
//
//     polyominoes(8).length;
//
// Get an array consisting of a single representation for each of them:
//
//     polyominoes(8).map(p => p.representation);

File: polyominoes.py
# A representation is a set of points corresponding to the location of each of
# the corresponding polyominoes constituent squares. A representation is
# normalized if the co-ordinates of all of its points are non-negative, at
# least one x co-ordinate is 0 and at least one y co-ordinate is 0.
def normalize(representation):
    dx = min(x for x, _ in representation)
    dy = min(y for _, y in representation)
    return frozenset((x - dx, y - dy) for x, y in representation)

# We'll store a polyomino as a set of the normalized representations of all its
# rotations and reflections.
def newPolyomino(representation):
    return frozenset(normalize(r) for r in [
        {( x,  y) for x, y in representation}, # identity
        {(-y,  x) for x, y in representation}, # rotate pi/2
        {(-x, -y) for x, y in representation}, # rotate pi
        {( y, -x) for x, y in representation}, # rotate 3pi/2
        {(-x,  y) for x, y in representation}, # reflect in x axis
        {( x, -y) for x, y in representation}, # reflect in y axis
        {( y,  x) for x, y in representation}, # reflect in y = x
        {(-y, -x) for x, y in representation}, # reflect in y = -x
    ])

# It's useful to be able to get a single representation for a given polyomino.
def represent(polyomino):
    return set(next(iter(polyomino)))

# We can build (n + 1)-ominoes by adding squares to n-ominoes
def grow(representation):
    return (newPolyomino(representation.union({s})) for s in {
        (x + dx, y + dy)
        for x, y in representation
        for dx, dy in [(1, 0), (0, 1), (-1, 0), (0, -1)]
        if (x + dx, y + dy) not in representation
    })

# This is everything we need to generate the set of all n-ominoes...
def polyominoes(n):
    ominoes = {1: {newPolyomino({(0, 0)})}}
    for i in range(1, n):
        ominoes[i + 1] = {a for b in ominoes[i] for a in grow(represent(b))}
    return [represent(p) for p in ominoes[n]]
